### DISCUSSION TITLE
What are the build instructions?

### DISCUSSION THREAD

**Discussion by GNO21** [yermak/AudioBookConverter#632] [Q&A]
What are the current build instructions to build this from source?

I've tried roughly following https://github.com/yermak/AudioBookConverter/wiki/Building-from-source with windows syntax, appropriate paths, and using the windows bat file.

Update: 
I managed to get it to work.
This assumes you have maven installed, along with the java sdk and javafx-jmods (though I think these files are now included, but still need their path referenced)

From a clean github clone: (git clone https://github.com/yermak/AudioBookConverter.git)

For windows, use a cmd shell:
1. Edit the pom.xml to have the appropriate version number for what you are doing
2. From the Audioconverter directory, run "mvn clean package --no-transfer-progress"
3. From the Audioconverter directory,run "build\windows\package.bat <path to java jdk> <path to javafx-jmods> (version that matches the om file)

Editing the POM file to set the version to match was one issue I had.
Using the git clone vs the source zip was the other.

**Comment by sgmoore** [comment]
This was helpful, but I still had issues getting this to build and I think this was because I had some incorrect versions of some programs already installed, so this is how I got my build working in **Windows SandBox** which means I am starting from a clean windows install. 

It was based on the contents of **build_binaries.yml** and I would recommend you check it for changes. 

Note The build process generates **a lot** or warnings and errors which is to be expected and doesn't stop your program from building.

### Install choco 

Open Powershell as admin and run the following command

`Set-ExecutionPolicy Bypass -Scope Process -Force; [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072; iex ((New-Object System.Net.WebClient).DownloadString('https://community.chocolatey.org/install.ps1'))
`

**Install the dependencies** 

`choco install git maven wget 7zip `


**Open a new Command prompt (NOT Powershell)**

**Clone the repo or your forked clone, e.g.**

```
md \source
cd \source
git clone https://github.com/yermak/AudioBookConverter
cd AudioBookConverter
start https://github.com/yermak/AudioBookConverter/releases 
``` 


The latter open the releases webpage so you can find the latest version number 
set the environment variable APP_VERSION to this version (or else a new version number) e.g.

`set APP_VERSION=6.5.2`

**Set the version** 

The replace task tokens replaces the /#{APP_VERSION}# placeholder with the correct version and I have replicated this using the sed that comes with Git .

```
"\Program Files\Git\usr\bin\sed.exe" -i 's/#{APP_VERSION}#/%APP_VERSION%/g' version.txt pom.xml src/main/java/uk/yermak/audiobookconverter/Version.java
type version.txt
rem verify it has changed correctly
```

the repo ships with javafx and mods which may depend on the version of Java, so again I have downloaded the version used by the  build_binaries.yml file and used that to build the package

```
wget https://download.java.net/openjdk/jdk21/ri/openjdk-21+35_windows-x64_bin.zip
7z.exe x openjdk-21+35_windows-x64_bin.zip
set JAVA_HOME=jdk-21
SET JAVAFX_JMODS=javafx/jmods/windows
mvn package
call build/windows/package.bat "%JAVA_HOME%" %JAVAFX_JMODS% "%APP_VERSION%"
```
Note the version section of the last command is different from the current yml file.

_This does not build the MSI file, so ignore the messages complaining about it._

**Run the compiled program** 

`target\image\AudioBookConverter\AudioBookConverter.exe`
